File: client-scalaz/src/main/scala/org/openapitools/client/api/Filter.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import Filter._

case class Filter (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
application: Option[Application],
name: Option[String],
description: Option[String],
parent: Option[Filter],
children: Option[List[Filter]],
display: Option[String],
externalId: Option[String],
externalType: Option[String],
tree: Option[String])

object Filter {
  import DateTimeCodecs._

  implicit val FilterCodecJson: CodecJson[Filter] = CodecJson.derive[Filter]
  implicit val FilterDecoder: EntityDecoder[Filter] = jsonOf[Filter]
  implicit val FilterEncoder: EntityEncoder[Filter] = jsonEncoderOf[Filter]
}

File: client-clojure/src/sirqul_io_t_platform/specs/filter.clj
(ns sirqul-io-t-platform.specs.filter
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.application :refer :all]
            [sirqul-io-t-platform.specs.filter :refer :all]
            [sirqul-io-t-platform.specs.filter :refer :all]
            )
  (:import (java.io File)))


(def filter-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :application) application-spec
   (ds/opt :name) string?
   (ds/opt :description) string?
   (ds/opt :parent) filter-spec
   (ds/opt :children) (s/coll-of filter-spec)
   (ds/opt :display) string?
   (ds/opt :externalId) string?
   (ds/opt :externalType) string?
   (ds/opt :tree) string?
   })

(def filter-spec
  (ds/spec
    {:name ::filter
     :spec filter-data}))
